src/clojure/nrepl/socket.clj

```diff
@@ -0,0 +1,222 @@
+(ns nrepl.socket
+  "Compatibility layer for java.io vs java.nio sockets to allow an
+  incremental transition to nio, since the JDK's filesystem sockets
+  don't support the java.io socket interface, and we can't use the
+  compatibility layer for bidirectional read and write:
+  https://bugs.openjdk.java.net/browse/JDK-4509080."
+  (:require
+   [clojure.java.io :as io]
+   [nrepl.misc :refer [log]])
+  (:import
+   (clojure.lang Reflector)
+   (java.io BufferedInputStream BufferedOutputStream File)
+   (java.net InetSocketAddress ServerSocket Socket StandardProtocolFamily URI)
+   (java.nio ByteBuffer)
+   (java.nio.file Path)
+   (java.nio.channels Channels ClosedChannelException
+                      ServerSocketChannel SocketChannel)))
+
+(def orig-warn-on-reflection *warn-on-reflection*)
+
+(defmacro find-class [full-path]
+  `(try
+     (import ~full-path)
+     (catch ClassNotFoundException ex#
+       nil)))
+
+;;; InetSockets (TCP)
+
+(defn inet-socket [bind port]
+  (let [port (or port 0)
+        addr (fn [^String bind ^Integer port] (InetSocketAddress. bind port))
+        ;; We fallback to 127.0.0.1 instead of to localhost to avoid
+        ;; a dependency on the order of ipv4 and ipv6 records for
+        ;; localhost in /etc/hosts
+        bind (or bind "127.0.0.1")]
+    (doto (ServerSocket.)
+      (.setReuseAddress true)
+      (.bind (addr bind port)))))
+
+;; Unix domain sockets
+
+(def ^Class junixsocket-address-class
+  (find-class 'org.newsclub.net.unix.AFUNIXSocketAddress))
+
+(def ^Class junixsocket-server-class
+  (find-class 'org.newsclub.net.unix.AFUNIXServerSocket))
+
+(def ^Class jdk-unix-address-class
+  (find-class 'java.net.UnixDomainSocketAddress))
+
+(def ^Class jdk-unix-server-class
+  (find-class 'java.nio.channels.ServerSocketChannel))
+
+(def ^:private test-junixsocket?
+  ;; Make it possible to test junixsocket even when JDK >= 16
+  (= "true" (System/getProperty "nrepl.test.junixsocket")))
+
+(def ^:const unix-domain-flavor
```

Comment:
:const is not a feature that is officially described in https://clojure.org/ so I'd just like it a vanilla def.

Savvy users can use direct linking which is cleaner and more generalized

Reply:
If it matters, I didn't add this for performance, I added it as a signal that other code may (and I think right now does) assume that the value will never change, but happy to remove it if it's undesirable, or ineffective.

Reply:
I'd say that Clojure programmers have the general assumption that vars never change - otherwise it's signaled by the use of a reference type (atoms, etc)

So I'd vote on removing it, FWIW :)